eclipse-theia/theia
# Enable localization of monaco-editor

#### What it does

Fixes one of the open issues from #9538: Enables the localization of the monaco-editor using vscode language packs.

Monaco is using a quite complicated localization mechanic. We are able to inject our custom translation into monaco using the `loadBundle` function. The callback receives a localization bundle in the form of `Record<string, string[]>` which is build from the language pack (which uses the form of `Record<string, Record<string, string>>`). The bulk of the work translating between these formats is done using a script (`translate-monaco.js`).

The script tries to match the indices of the currently used monaco translation (array based) to the provided language pack. This works mostly fine, aside from two special cases (`keybindingLabels` and `findWidget`). There is a special duplicating logic inside of the script, that needs to be used due to duplicated entries inside of the monaco `nls.js` files.

There would be an alternative approach to translating monaco: Update the `@theia/monaco-editor-core` package to natively include the fitting translations. That would alleviate the need for all of this. However, the drawback of that method would be that only selected localizations would be available. For example, the `monaco-editor` package on npm does not provide a turkish translation. Using the turkish language pack with the current proposed approach translates monaco into turkish perfectly. We are able to provide arbitrary translations for monaco using language packs this way.

The current proposal comes with another drawback: The provided `monaco-nls.json` has to be updated with every monaco update. The scripts should be able to deal with that, but there may be need for small adjustments in the script itself. Another possible drawback is that future language pack versions might not be compatible with the old monaco-editor translations. I couldn't observe any issues with that using a language version from vscode `1.60.4`, so we don't have to worry about that for quite some time.

#### How to test

1. Download a language pack of your choice from the ovsx-registry (version `1.53.2` or newer).
2. Use the configure language command to change to that language.
3. Confirm that the monaco related parts of Theia have been correctly translated:
    - The editor context menu
    - The find widget
    - The `Selection` main menu entry and its associated commands
4. You can also try the script itself using the instructions provided in the script itself.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
